telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/bot/handlers/impl/financehandlers/AddFinancesModelHandler.kt

```diff
@@ -0,0 +1,121 @@
+import com.ajaxproject.telegrambot.bot.enums.TextPropertyName.SUCCESSFUL_ADD_FINANCE
+import com.ajaxproject.telegrambot.bot.handlers.UserRequestHandler
+import com.ajaxproject.financemodels.models.MongoFinance
+import com.ajaxproject.financemodels.enums.Finance.INCOME
+import com.ajaxproject.financemodels.enums.Finance.EXPENSE
+import com.ajaxproject.telegrambot.bot.enums.Buttons.BACK_TO_MENU_BUTTON
+import com.ajaxproject.telegrambot.bot.enums.Buttons.ADD_FINANCE_AGAIN
+import com.ajaxproject.telegrambot.bot.service.FinanceRequestNatsService
+import com.ajaxproject.telegrambot.bot.service.TelegramService
+import com.ajaxproject.telegrambot.bot.service.TextService
+import com.ajaxproject.telegrambot.bot.service.UserSessionService
+import com.ajaxproject.telegrambot.bot.service.isTextMessage
+import com.ajaxproject.telegrambot.bot.service.updatemodels.UpdateRequest
+import com.ajaxproject.telegrambot.bot.utils.KeyboardUtils
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.core.scheduler.Schedulers
+import reactor.kotlin.core.publisher.switchIfEmpty
+import java.util.*
+
+@Component
+class AddFinancesModelHandler(
+    private val telegramService: TelegramService,
+    private val financeRequestNatsService: FinanceRequestNatsService,
+    private val textService: TextService,
+    private val userSessionService: UserSessionService,
+) : UserRequestHandler {
+
+    override fun isApplicable(request: UpdateRequest): Boolean {
+        return WAITING_FOR_ADD_FINANCE == request.updateSession.state &&
+                request.update.isTextMessage()
+    }
+
+    override fun handle(dispatchRequest: UpdateRequest) {
+        val financeData = dispatchRequest.update.message.text
+        val chatId = dispatchRequest.chatId
+
+        Mono.just(financeData)
+            .filter { it.matches(Regex("[+-]\\d+ [^\\n\\r]+\$")) }
+            .switchIfEmpty {
+                Mono.fromSupplier { sendMessageIfDataIsNotCorrect(chatId) }
+                    .subscribeOn(Schedulers.boundedElastic())
+                    .then(Mono.empty())
+            }
+            .flatMap {
+                sendRequestToCreateFinance(dispatchRequest, financeData)
+                    .subscribeOn(Schedulers.boundedElastic())
+            }
+            .flatMap {
+                Mono.fromSupplier {
+                    sendMessageIfSuccessfulCreateFinance(chatId)
+                }.subscribeOn(Schedulers.boundedElastic())
+            }
+            .flatMap {
+                Mono.fromSupplier {
+                    dispatchRequest.updateSession.apply { state = CONVERSATION_STARTED }
+                }
+            }
+            .flatMap { session ->
+                userSessionService.saveSession(dispatchRequest.chatId, session)
+                Mono.just(session)
+            }
+            .subscribe()
+    }
+
+    private fun sendMessageIfDataIsNotCorrect(chatId: Long) {
+        telegramService.sendMessage(
```

Comment:
looks like you need to make telegramService reactive, because you have too many boilerplate code to work with it

Reply:
I will rewrite logic in all my handlers